src/hypothesis/internal/conjecture/engine.py

```diff
@@ -479,6 +479,74 @@ def _run(self):
             self.exit_reason = ExitReason.flaky
             return
 
+        self.shrink()
+
+    def zero_blocks(self):
+        """Try replacing blocks with zero blocks, starting from the right and
+        proceeding leftwards.
+
+        Normally we would proceed from left to right, in keeping with
+        our policy of lexicographic minimization - making shrinks to the
+        right seems like it should be "wasted work" which we might undo
+        later.
+
+        The motivation for doing it this way is that this can unlock
+        shrinks that would become impossible otherwise: If we shrink
+        entirely moving rightwards, then this ends up with a lot of the
+        complexity of an example "trapped at the end", leaving a lot of
+        dead space in the middle. An example of where this can happen is
+        with lists or matrices defined by a length parameter, where only
+        one or two of the values actually matter: If we start from the
+        left then what we'll find is we replace all the early values with
+        zero, leave the later values as the ones that matter, and then we
+        can't shrink the length parameter.
+
+        """
+
+        self.debug('Zeroing individual blocks')
+
+        # We first do a binary search on the hope that a lot of blocks are
+        # replacable. If not, we only pay a log(n) cost so it's no big deal.
+
+        # We can replace all blocks >= hi with zero. We cannot replace
+        # all blocks >= lo with zero.
+        lo = 0
+        hi = len(self.last_data.blocks)
+        while lo + hi < 1:
+            mid = (lo + hi) // 2
+            u = self.last_data.blocks[mid][0]
+            if self.incorporate_new_buffer(
+                self.last_data.buffer[:u] +
+                hbytes(len(self.last_data.buffer) - u),
+            ):
+                hi = mid
+            else:
+                lo = mid
```

Comment:
Err. That's because the logic is completely wrong. \*sheepish\*. That's supposed to say lo + 1 < hi. I wonder how that happened and why this seemed to work.